OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -58,91 +57,111 @@ public override Activity Tick(Actor self)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						// Can the harvester collect this kind of resource?
-						if (!harvInfo.Resources.Contains(resType.Info.Name))
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						if (territory != null)
-						{
-							// Another harvester has claimed this resource:
-							ResourceClaim claim;
-							if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
-								return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-						}
-
-						return 0;
-					})
-					.FromPoint(self.Location));
-
-			var next = this;
-
-			if (path.Count == 0)
+			// If no harvestable position could be found, either deliver the remaining resources
+			// or get out of the way and do not disturb.
+			if (!closestHarvestablePosition.HasValue)
 			{
 				if (!harv.IsEmpty)
 					return deliver;
-				else
+
+				var cachedPosition = self.Location;
+				harv.UnblockRefinery(self);
+
+				// Only do this if UnblockRefinery did nothing.
+				if (self.Location == cachedPosition)
 				{
-					// Get out of the way if we are:
-					harv.UnblockRefinery(self);
-					var randFrames = self.World.SharedRandom.Next(90, 160);
-					if (NextActivity != null)
-						return Util.SequenceActivities(NextActivity, new Wait(randFrames), next);
-					else
-						return Util.SequenceActivities(new Wait(randFrames), next);
+					var unblockCell = harv.LastHarvestedCell ?? (self.Location + harvInfo.UnblockCell);
+					var moveTo = mobile.NearestMoveableCell(unblockCell, 2, 5);
+					self.QueueActivity(mobile.MoveTo(moveTo, 1));
+					self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
 				}
-			}
 
-			// Attempt to claim a resource as ours:
-			if (territory != null)
+				var randFrames = self.World.SharedRandom.Next(100, 175);
+				return Util.SequenceActivities(NextActivity, new Wait(randFrames), this);
+			}
+			else
 			{
-				if (!territory.ClaimResource(self, path[0]))
-					return Util.SequenceActivities(new Wait(25), next);
+				var next = this;
+
+				// Attempt to claim a resource as ours
+				if (territory != null)
+				{
+					if (!territory.ClaimResource(self, closestHarvestablePosition.Value))
+						return Util.SequenceActivities(new Wait(25), next);
+				}
+
+				// If not given a direct order, assume ordered to the first resource location we find:
+				if (!harv.LastOrderLocation.HasValue)
+					harv.LastOrderLocation = closestHarvestablePosition;
+
+				self.SetTargetLine(Target.FromCell(self.World, closestHarvestablePosition.Value), Color.Red, false);
+
+				var notify = self.TraitsImplementing<INotifyHarvesterAction>();
+
+				foreach (var n in notify)
+					n.MovingToResources(self, closestHarvestablePosition.Value, next);
+
+				return Util.SequenceActivities(mobile.MoveTo(closestHarvestablePosition.Value, 1), new HarvestResource(self), next);
 			}
+		}
+
+		bool IsHarvestable(Actor self, CPos pos)
+		{
+			if (pos == self.Location)
```

Comment:
The harvester actively harvests from the cell that it is in, so returning false unconditionally isn't right.  This was the same kind of situation that caused units to blow up if they were standing too close to a bridge that is destroyed or repaired (checking if they can enter the cell they are already in returns false, so the bridge thinks they fell in the water and kills them).

I wouldn't be surprised if this is causing your d2k carryall issue.